pkg/metrics/api/api.go

```diff
@@ -0,0 +1,456 @@
+)
+
+var (
+	// FileCategories is a map of asset names to the categories of the metrics that it belongs to.
+	FileCategories = make(map[string]string)
+	// metricRegistry holds all the MetricBuilder objects for all the metrics that have been designed
+	// and ready for extraction from the installer.
+	metricRegistry = make(map[string]builder.MetricBuilder)
+
+	// CreateCommandMetricName is the map of the targets to the respective metric names. Installer can
+	// look this up for the correct metric name given a target.
+	CreateCommandMetricName = make(map[string]string)
+
+	// optsRegistry holds all the information that every metric needs to create a metric builder.
+	optsRegistry = make(map[string]builder.MetricOpts)
+
+	// CurrentInvocationContext is used to store the current Invocation metric name used for building context.
+	// Invocation data information is gathered from different parts of the installer and it gets hard for the
+	// installer to keep track of the current command and target that is being run. Hence, the current invocation
+	// metric name is stored here in this variable.
+	CurrentInvocationContext string
+
+	// enableMetrics holds the user preference for sending metrics. If the user chooses not to send, this will hold
+	// the value false.
+	enableMetrics bool
+
+	// prometheusURL is the standard URL to send metrics to. The user can override this information to any URL they choose.
+	prometheusURL = "https://localhost:9092/"
+
+	// pushClient holds the URL, jobname and the http client information to send metrics.
+	pushClient pushclient.PushClient
+)
+
+// Initialize holds all the initial setup information like creating all the metrics that are allowed to
+// be extracted, checking for user input about disabling metrics, URL and for creating the Push Client.
+// Must be run before the metircs are collected.
+func Initialize() {
+	enableMetrics = false
+
+	// if _, ok := os.LookupEnv("OPENSHIFT_INSTALL_DISABLE_METRICS"); ok {
+	// 	enableMetrics = true
+	// }
+
+	if value, ok := os.LookupEnv("OPENSHIFT_INSTALL_METRICS_ENDPOINT"); ok {
+		prometheusURL = value
+		enableMetrics = true
+	}
+	initializeModificationCategories()
+	initializeCreateCommandMetricName()
+	initializeOptsForMetrics()
+
+	pushClient = pushclient.PushClient{url: prometheusURL, Client: &http.Client{}, JobName: "openshift_installer_metrics"}
+}
+
+// initializeCreateCommandMetricName populates the CreateCommandMetricName to make it easy for the installer to get the metric name
+// to populate for the given target.
+func initializeCreateCommandMetricName() {
+	CreateCommandMetricName["cluster"] = ClusterInstallationCreateJobName
+	CreateCommandMetricName["ignition-configs"] = ClusterInstallationIgnitionJobName
+	CreateCommandMetricName["install-config"] = ClusterInstallationInstallConfigJobName
+	CreateCommandMetricName["manifests"] = ClusterInstallationManifestsJobName
```

Comment:
Why do we need this? Metric names are exported so can't the calling code just use the appropriate metric name?

Reply:
I still need to find the metric name for the given target. I thought I can either write it in the installer create command or here for the installer to call and the latter seems to make sense to me.